adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,546 @@
+        LogError("Invalid parameter");
+        result = __FAILURE__;
+    }
+    else if (tlsio_instance->state != TLSIO_APPLEIOS_STATE_OPEN)
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_058: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, TLSIO_APPLEIOS_STATE_OPENING, TLSIO_APPLEIOS_STATE_CLOSING, or TLSIO_APPLEIOS_STATE_CLOSED, the tlsio_appleios_send shall call the on_send_complete with IO_SEND_ERROR, and return _LINE_. ]*/
+        LogError("TLS is not ready to send data");
+        result = __FAILURE__;
+    }
+    else
+    {
+        CFIndex send_result;
+        size_t send_size = size;
+        const uint8_t* runBuffer = (const uint8_t *)buffer;
+        result = __FAILURE__;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_055: [ if the ssl was not able to send all data in the buffer, the tlsio_appleios_send shall call the ssl again to send the remaining bytes. ]*/
+        while (send_size > 0)
+        {
+			send_result = CFWriteStreamWrite(tlsio_instance->sockWrite, buffer, size);
+			
+			if (send_result == -1)
+            {
+                /* Codes_SRS_TLSIO_APPLEIOS_32_056: [ if the ssl was not able to send any byte in the buffer, the tlsio_appleios_send shall call the on_send_complete with IO_SEND_ERROR, and return _LINE_. ]*/
+                LogError("TLS failed sending data");
+                /* Codes_SRS_TLSIO_APPLEIOS_32_053: [ The tlsio_appleios_send shall use the provided on_io_send_complete callback function address. ]*/
+                /* Codes_SRS_TLSIO_APPLEIOS_32_054: [ The tlsio_appleios_send shall use the provided on_io_send_complete_context handle. ]*/
+                if (on_send_complete != NULL)
+                {
+                    /* Codes_SRS_TLSIO_APPLEIOS_32_003: [ The tlsio_appleios shall report the send operation status using the IO_SEND_RESULT enumerator defined in the `xio.h`. ]*/
+                    on_send_complete(callback_context, IO_SEND_ERROR);
+                }
+                send_size = 0;
+            }
+            else if (send_result >= send_size) /* Transmit it all. */
+            {
+                /* Codes_SRS_TLSIO_APPLEIOS_32_057: [ if the ssl finish to send all bytes in the buffer, the tlsio_appleios_send shall call the on_send_complete with IO_SEND_OK, and return 0 ]*/
+                /* Codes_SRS_TLSIO_APPLEIOS_32_053: [ The tlsio_appleios_send shall use the provided on_io_send_complete callback function address. ]*/
+                /* Codes_SRS_TLSIO_APPLEIOS_32_054: [ The tlsio_appleios_send shall use the provided on_io_send_complete_context handle. ]*/
+                if (on_send_complete != NULL)
+                {
+                    /* Codes_SRS_TLSIO_APPLEIOS_32_003: [ The tlsio_appleios shall report the send operation status using the IO_SEND_RESULT enumerator defined in the `xio.h`. ]*/
+                    on_send_complete(callback_context, IO_SEND_OK);
+                }
```

Comment:
Callbacks should not be called in case send_async failed immediately. The caller is already informed of the error and the convention is that the callback is not expected if send has failed.

Reply:
This is a success callback

---
In reply to: [139067554](https://github.com/Azure/azure-c-shared-utility/pull/109#discussion_r139067554) [](ancestors = 139067554)

Reply:
Oh I see, cool.

---
In reply to: [139553012](https://github.com/Azure/azure-c-shared-utility/pull/109#discussion_r139553012) [](ancestors = 139553012,139067554)